torchao/kv_cache.py

```diff
@@ -0,0 +1,415 @@
+import torch
+import torch.nn as nn
+import functools
+from typing import List, Tuple, Optional, Dict, Any
+import copy
+
+HANDLED_FUNCTIONS = {}
+
+
+class PagedTensor(object):
+    def __init__(
+        self,
+        cache: torch.Tensor,
+        block_tables: torch.Tensor,
+        context_lens: torch.Tensor,
+    ):
+        self.block_tables = block_tables
+        self.cache = cache
+        self.context_lens = context_lens
+
+    def __repr__(self):
+        return f"PagedTensor({self.cache.shape})"
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        if func not in HANDLED_FUNCTIONS or not all(
+            issubclass(t, (torch.Tensor, PagedTensor)) for t in types
+        ):
+            return NotImplementedError(
+                "{} is not supported by PagedTensor".format(func)
+            )
+        return HANDLED_FUNCTIONS[func](*args, **kwargs)
+
+
+def implements(torch_function):
+    """Register a torch function override for PagedTensor"""
+
+    def decorator(func):
+        functools.update_wrapper(func, torch_function)
+        HANDLED_FUNCTIONS[torch_function] = func
+        return func
+
+    return decorator
+
+
+@implements(torch.nn.functional.scaled_dot_product_attention)
+def scaled_dot_product_attention(
+    input, key_tensor, value_tensor, attn_mask=None, scale=None
+):
+    query = input
+    key_cache = key_tensor.cache
+    value_cache = value_tensor.cache
+    num_kv_head = key_cache.size(1)
+    num_queries_per_kv = query.size(1) // num_kv_head
+    block_size = key_cache.size(2)
+    block_tables = key_tensor.block_tables
+    head_mapping = torch.repeat_interleave(
+        torch.arange(num_kv_head, dtype=torch.int32, device="cpu"), num_queries_per_kv
+    )
```

Comment:
Can we do this inside the `paged_attention` c++ kernel so that we don't need to pass this `head_mapping` arg to it? This simplifies the kernel interface.